src/v/cluster/rm_stm.cc

```diff
@@ -0,0 +1,1054 @@
+    // to invalidate the post processing
+    auto term = _mem_state.term;
+
+    auto r = co_await _c->replicate(
+      _mem_state.term,
+      std::move(br),
+      raft::replicate_options(raft::consistency_level::leader_ack));
+    if (!r) {
+        if (_mem_state.estimated.contains(bid.pid)) {
+            // an error during replication, preventin tx from progress
+            _mem_state.expected.erase(bid.pid);
+        }
+        co_return kafka::error_code::unknown_server_error;
+    }
+
+    auto replicated = r.value();
+
+    if (_mem_state.term != term) {
+        // mem state already changed no need to clean it
+        co_return replicated;
+    }
+
+    auto last_offset = model::offset(replicated.last_offset());
+    if (!_mem_state.tx_start.contains(bid.pid)) {
+        auto base_offset = model::offset(
+          last_offset() - (bid.record_count - 1));
+        _mem_state.tx_start.emplace(bid.pid, base_offset);
+        _mem_state.tx_starts.insert(base_offset);
+        _mem_state.estimated.erase(bid.pid);
+    }
+    co_return replicated;
+}
+
+ss::future<checked<raft::replicate_result, kafka::error_code>>
+rm_stm::replicate_seq(
+  model::batch_identity bid,
+  model::record_batch_reader br,
+  raft::replicate_options opts) {
+    auto is_ready = co_await sync(_sync_timeout);
+    if (!is_ready) {
+        co_return kafka::error_code::not_leader_for_partition;
+    }
+    if (_mem_state.term != _insync_term) {
+        _mem_state = mem_state{.term = _insync_term};
+    }
+    if (!check_seq(bid)) {
+        co_return kafka::error_code::out_of_order_sequence_number;
+    }
+    auto r = co_await _c->replicate(_insync_term, std::move(br), opts);
+    if (!r) {
+        co_return kafka::error_code::unknown_server_error;
+    }
+    co_return r.value();
+}
+
+model::offset rm_stm::last_stable_offset() {
+    auto first_tx_start = model::offset(std::numeric_limits<int64_t>::max());
+
+    if (!_log_state.ongoing_set.empty()) {
+        first_tx_start = *_log_state.ongoing_set.begin();
+    }
+
+    if (!_mem_state.tx_starts.empty()) {
+        first_tx_start = std::min(
+          first_tx_start, *_mem_state.tx_starts.begin());
+    }
+
+    for (auto& entry : _mem_state.estimated) {
+        first_tx_start = std::min(first_tx_start, entry.second);
+    }
+
+    auto last_visible_index = _c->last_visible_index();
+    if (first_tx_start <= last_visible_index) {
+        return first_tx_start;
+    }
+
+    return raft::details::next_offset(last_visible_index);
+}
+
+ss::future<std::vector<rm_stm::tx_range>>
+rm_stm::aborted_transactions(model::offset from, model::offset to) {
+    std::vector<rm_stm::tx_range> result;
+    for (auto& range : _log_state.aborted) {
+        if (range.second.last < from) {
+            continue;
+        }
+        if (range.second.first > to) {
+            continue;
+        }
+        result.push_back(range.second);
+    }
+    co_return result;
```

Comment:
it doesn't need now but when we start storing aborted transactions in an on disk file instead of memory then the method will become async, it's cheaper to treat this function as async now then change the dependancies for sync to async later